File: app/lib/models/filter_registry.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter/services.dart';
import 'package:path/path.dart' as path;

import 'filter_schema.dart';

/// Registry for all available filter schemas.
///
/// Loads built-in filters from assets and user filters from the filesystem.
class FilterRegistry {
  static final FilterRegistry instance = FilterRegistry._();
  FilterRegistry._();

  final Map<String, FilterSchema> _filters = {};
  final List<String> _loadOrder = [];
  bool _isLoaded = false;

  /// Whether the registry has been initialized.
  bool get isLoaded => _isLoaded;

  /// Get all registered filters in order.
  List<FilterSchema> get filters =>
      _loadOrder.map((id) => _filters[id]!).toList();

  /// Get all filters sorted by their order property.
  List<FilterSchema> get orderedFilters =>
      filters.toList()..sort((a, b) => a.order.compareTo(b.order));

  /// Get filters by category.
  List<FilterSchema> getByCategory(String category) =>
      filters.where((f) => f.category == category).toList();

  /// Get a filter by ID.
  FilterSchema? get(String id) => _filters[id];

  /// Check if a filter exists.
  bool has(String id) => _filters.containsKey(id);

  /// Initialize the registry by loading all filters.
  Future<void> initialize() async {
    if (_isLoaded) return;

    // Load built-in filters first
    await _loadBuiltInFilters();

    // Then load user filters (can override built-in)
    await _loadUserFilters();

    _isLoaded = true;
  }

  /// Reload all filters (useful after user adds new filters).
  Future<void> reload() async {
    _filters.clear();
    _loadOrder.clear();
    _isLoaded = false;
    await initialize();
  }

  /// Load built-in filters from app assets.
  Future<void> _loadBuiltInFilters() async {
    try {
      // Load the manifest to get list of filter files
      final manifestJson =
          await rootBundle.loadString('assets/filters/manifest.json');
      final manifest = json.decode(manifestJson) as List;

      for (final filename in manifest) {
        try {
          final schemaJson =
              await rootBundle.loadString('assets/filters/core/$filename');
          final schema = FilterSchema.fromJson(json.decode(schemaJson));
          schema.source = 'builtin';
          _register(schema);
        } catch (e) {
          print('Failed to load built-in filter $filename: $e');
        }
      }
    } catch (e) {
      // Manifest doesn't exist yet, that's fine during development
      print('No filter manifest found: $e');
    }
  }

  /// Load user filters from the user's filter directory.
  Future<void> _loadUserFilters() async {
    final userDir = await _getUserFilterDirectory();
    if (userDir == null || !await userDir.exists()) return;

    await for (final entity in userDir.list(recursive: true)) {
      if (entity is File && entity.path.endsWith('.json')) {
        try {
          final content = await entity.readAsString();
          final schema = FilterSchema.fromJson(json.decode(content));
          schema.source = 'user';
          _register(schema);
        } catch (e) {
          print('Failed to load user filter ${entity.path}: $e');
        }
      }
    }
  }

  /// Get the user filter directory path.
  Future<Directory?> _getUserFilterDirectory() async {
    String? home;
    if (Platform.isWindows) {
      home = Platform.environment['USERPROFILE'];
    } else {
      home = Platform.environment['HOME'];
    }

    if (home == null) return null;

    return Directory(path.join(home, '.vapourbox', 'filters'));
  }

  /// Register a filter schema.
  void _register(FilterSchema schema) {
    // Remove from load order if already exists (for overrides)
    _loadOrder.remove(schema.id);

    _filters[schema.id] = schema;
    _loadOrder.add(schema.id);
  }

  /// Register a filter from JSON (useful for testing).
  void registerFromJson(Map<String, dynamic> json, {String source = 'dynamic'}) {
    final schema = FilterSchema.fromJson(json);
    schema.source = source;
    _register(schema);
  }

  /// Validate that all required dependencies are available.
  Future<Map<String, List<String>>> validateDependencies() async {
    final missing = <String, List<String>>{};

    for (final filter in filters) {
      final deps = filter.dependencies;
      if (deps == null) continue;

      final filterMissing = <String>[];

      // Check VS plugins
      if (deps.vsPlugins != null) {
        for (final plugin in deps.vsPlugins!) {
          // This would need to be implemented based on platform
          // For now, we'll assume they're available
        }
      }

      if (filterMissing.isNotEmpty) {
        missing[filter.id] = filterMissing;
      }
    }

    return missing;
  }
}

File: worker/src/filter_registry.rs
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};

use anyhow::{Context, Result};

use crate::filter_schema::FilterSchema;

/// Registry for all available filter schemas.
///
/// Loads built-in filters from the schemas directory and user filters from
/// the user's config directory.
pub struct FilterRegistry {
    filters: HashMap<String, FilterSchema>,
    load_order: Vec<String>,
}

impl FilterRegistry {
    /// Create a new empty registry.
    pub fn new() -> Self {
        Self {
            filters: HashMap::new(),
            load_order: Vec::new(),
        }
    }

    /// Load all available filters.
    pub fn load_all(&mut self, schemas_dir: &Path) -> Result<()> {
        // Load built-in filters
        self.load_from_directory(schemas_dir, "builtin")?;

        // Load user filters
        if let Some(user_dir) = Self::get_user_filter_directory() {
            if user_dir.exists() {
                self.load_from_directory(&user_dir, "user")?;
            }
        }

        Ok(())
    }

    /// Load filters from a directory.
    pub fn load_from_directory(&mut self, dir: &Path, source: &str) -> Result<()> {
        if !dir.exists() {
            return Ok(());
        }

        // Load from core/ subdirectory if it exists
        let core_dir = dir.join("core");
        let search_dir = if core_dir.exists() { &core_dir } else { dir };

        for entry in fs::read_dir(search_dir)? {
            let entry = entry?;
            let path = entry.path();

            if path.extension().map(|e| e == "json").unwrap_or(false) {
                if let Err(e) = self.load_from_file(&path, source) {
                    eprintln!("Warning: Failed to load filter schema from {:?}: {}", path, e);
                }
            }
        }

        Ok(())
    }

    /// Load a single filter from a file.
    pub fn load_from_file(&mut self, path: &Path, source: &str) -> Result<FilterSchema> {
        let content = fs::read_to_string(path)
            .with_context(|| format!("Failed to read filter schema: {:?}", path))?;

        let mut schema: FilterSchema = serde_json::from_str(&content)
            .with_context(|| format!("Failed to parse filter schema: {:?}", path))?;

        schema.source = source.to_string();
        self.register(schema.clone());

        Ok(schema)
    }

    /// Register a filter schema.
    pub fn register(&mut self, schema: FilterSchema) {
        // Remove from load order if already exists (for overrides)
        self.load_order.retain(|id| id != &schema.id);

        self.filters.insert(schema.id.clone(), schema);
        self.load_order.push(self.filters.keys().last().unwrap().clone());
    }

    /// Get all registered filters in load order.
    pub fn filters(&self) -> impl Iterator<Item = &FilterSchema> {
        self.load_order.iter().filter_map(|id| self.filters.get(id))
    }

    /// Get all filters sorted by their order property.
    pub fn ordered_filters(&self) -> Vec<&FilterSchema> {
        let mut filters: Vec<_> = self.filters.values().collect();
        filters.sort_by_key(|f| f.order);
        filters
    }

    /// Get a filter by ID.
    pub fn get(&self, id: &str) -> Option<&FilterSchema> {
        self.filters.get(id)
    }

    /// Check if a filter exists.
    pub fn has(&self, id: &str) -> bool {
        self.filters.contains_key(id)
    }

    /// Get the user filter directory path.
    fn get_user_filter_directory() -> Option<PathBuf> {
        #[cfg(windows)]
        {
            std::env::var("USERPROFILE")
                .ok()
                .map(|home| PathBuf::from(home).join(".vapourbox").join("filters"))
        }

        #[cfg(not(windows))]
        {
            std::env::var("HOME")
                .ok()
                .map(|home| PathBuf::from(home).join(".vapourbox").join("filters"))
        }
    }

    /// Validate that all required dependencies are available.
    pub fn validate_dependencies(&self, plugin_dir: &Path) -> HashMap<String, Vec<String>> {
        let mut missing = HashMap::new();

        for filter in self.filters.values() {
            if let Some(deps) = &filter.dependencies {
                let mut filter_missing = Vec::new();

                // Check VS plugins
                if let Some(vs_plugins) = &deps.vs_plugins {
                    for plugin in vs_plugins {
                        let plugin_path = plugin_dir.join(plugin);
                        if !plugin_path.exists() {
                            filter_missing.push(plugin.clone());
                        }
                    }
                }

                if !filter_missing.is_empty() {
                    missing.insert(filter.id.clone(), filter_missing);
                }
            }
        }

        missing
    }
}

impl Default for FilterRegistry {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn test_load_from_json() {
        let mut registry = FilterRegistry::new();

        let json = r#"{
            "id": "test_filter",
            "version": "1.0.0",
            "name": "Test Filter",
            "methods": [
                {
                    "id": "method1",
                    "name": "Method 1",
                    "function": "test.func",
                    "parameters": []
                }
            ],
            "parameters": {}
        }"#;

        let schema: FilterSchema = serde_json::from_str(json).unwrap();
        registry.register(schema);

        assert!(registry.has("test_filter"));
        assert_eq!(registry.get("test_filter").unwrap().name, "Test Filter");
    }

    #[test]
    fn test_load_from_directory() {
        let dir = tempdir().unwrap();
        let core_dir = dir.path().join("core");
        fs::create_dir(&core_dir).unwrap();

        let schema_json = r#"{
            "id": "file_filter",
            "version": "1.0.0",
            "name": "File Filter",
            "methods": [],
            "parameters": {}
        }"#;

        fs::write(core_dir.join("test.json"), schema_json).unwrap();

        let mut registry = FilterRegistry::new();
        registry.load_from_directory(dir.path(), "test").unwrap();

        assert!(registry.has("file_filter"));
    }
}
